plugins/algoltek-usb/fu-algoltek-usb-device.c

```diff
@@ -0,0 +1,514 @@
+{
+	g_autoptr(GByteArray) buf = fu_struct_algoltek_cmd_address_pkt_new();
+
+	fu_struct_algoltek_cmd_address_pkt_set_len(buf, 3);
+	fu_struct_algoltek_cmd_address_pkt_set_cmd(buf, FU_ALGOLTEK_CMD_ERS);
+	fu_struct_algoltek_cmd_address_pkt_set_checksum(buf, ~fu_sum8(buf->data, buf->len) + 1);
+
+	if (!algoltek_device_ctrl_transfer(self,
+					   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					   FU_ALGOLTEK_CMD_ERS,
+					   0,
+					   0,
+					   buf,
+					   buf->len,
+					   error)) {
+		g_prefix_error(error, "data clear failure: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_device_status_check(FuDevice *self, gpointer user_data, GError **error)
+{
+	guint16 update_status;
+	g_autoptr(GByteArray) update_status_array = g_byte_array_new();
+
+	update_status_array =
+	    algoltek_device_rdr(FU_ALGOLTEK_USB_DEVICE(self), AG_UPDATE_STATUS, error);
+	if (update_status_array == NULL)
+		return FALSE;
+
+	update_status = update_status_array->data[0];
+
+	switch (update_status) {
+	case AG_UPDATE_PASS:
+		break;
+	case AG_UPDATE_FAIL:
+	default:
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "update procedure is failed.");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+algoltek_device_wrf(FuAlgoltekUsbDevice *self,
+		    GBytes *blob_payload,
+		    FuProgress *progress,
+		    GError **error)
+{
+	const guint8 *fw_data = NULL;
+	gsize fw_data_size = 0;
+	guint16 value;
+	guint16 index;
+	g_autoptr(GByteArray) transfer_parameter = g_byte_array_new();
+	g_autoptr(GPtrArray) chunks_payload = NULL;
+
+	fw_data = g_bytes_get_data(blob_payload, &fw_data_size);
+	chunks_payload = fu_chunk_array_new(fw_data, fw_data_size, 0x0, 0x0, 64);
+
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, chunks_payload->len);
+
+	for (guint i = 0; i < chunks_payload->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks_payload, i);
+		g_autoptr(GByteArray) part_firmware = g_byte_array_new();
+
+		g_byte_array_append(part_firmware,
+				    fu_chunk_get_data(chk),
+				    fu_chunk_get_data_sz(chk));
+
+		fu_byte_array_set_size(transfer_parameter, 4, 0);
+		if ((i + 1) % 4 == 0)
+			transfer_parameter->data[0] = 1;
+		else
+			transfer_parameter->data[0] = 0;
+
+		fu_memwrite_uint24(transfer_parameter->data + 1,
+				   fu_chunk_get_address(chk),
+				   G_BIG_ENDIAN);
+
+		value = fu_memread_uint16(transfer_parameter->data, G_BIG_ENDIAN);
+		index = fu_memread_uint16(transfer_parameter->data + 2, G_BIG_ENDIAN);
+
+		if (!algoltek_device_ctrl_transfer(self,
+						   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+						   FU_ALGOLTEK_CMD_WRF,
+						   value,
+						   index,
+						   part_firmware,
+						   part_firmware->len,
+						   error)) {
+			g_prefix_error(error, "data write failure: ");
+			return FALSE;
+		}
+
+		if ((i + 1) % 4 == 0 || (i + 1) == chunks_payload->len) {
+			if (!fu_device_retry_full(FU_DEVICE(self),
+						  fu_algoltek_device_status_check,
```

Comment:
normally the function we call as a callback has a `_cb` suffix, e.g. `fu_algoltek_device_status_check_cb`

Reply:
We will check the code and add suffix `_cb` suffix to the callback function.

Reply:
Add the `_cb` suffix.